### DISCUSSION TITLE
How is external multithreading handled in twinBASIC

### DISCUSSION THREAD

**Discussion by bclothier** [twinbasic/twinbasic#824] [General]
> @bclothier I never overcame multi-threading issues with VB6 runtime in [VszLib](https://github.com/wqweto/VszLib) effort as some of the 7-zip formats performs I/O on separate threads (e.g. .rar) and these worker threads are **not** initialized for COM even (needed CoInitialize on callback entry).

_Originally posted by @wqweto in https://github.com/WaynePhillipsEA/twinbasic/discussions/808#discussioncomment-2672838_

(Starting a separate discussion to avoid straying from the discussion's original topic)

@WaynePhillipsEA, I put this to test and was successful extracting 3 different `.rar` archive files of different sizes. However, even with MTA, it appears that all access to the codebase is on a single thread. Can you confirm if the tB compiler will perform additional work to synchronize the access to the thread that owns the tB objects? 

I can't be 100% certain whether the workers on the 7z's library were in fact running on a different thread and whether the IDE is actually reporting new threads that might been spawned by external code.

**Comment by WaynePhillipsEA** [comment]
> Can you confirm if the tB compiler will perform additional work to synchronize the access to the thread that owns the tB objects?

No, tB does nothing special here.

> I can't be 100% certain whether the workers on the 7z's library were in fact running on a different thread and whether the IDE is actually reporting new threads that might been spawned by external code.

The IDE would only detect/report them if they break into the debugger.  So you can be running tB code on another thread unknowingly unless you hit a breakpoint or error.   You could log `App.ThreadID` to detect if that is happening.

**Reply by bclothier** [reply]
I did use win32's `GetCurrentThreadId` which seem to give the same result as the `App.ThreadId` and the debug messages report the same thread being used. Unless 7-zip changed how it handled multi-threaded I/O, it looks like it's running on a single thread in spite of reports of this not working on VB6's version.

**Reply by WaynePhillipsEA** [reply]
`App.ThreadId` is just a wrapper for `GetCurrentThreadId`, so that makes sense.  I'm not at all familiar with the 7-zip API, but my only question would be; did you test a reasonably large rar file? _(I'm guessing you probably did)_

**Reply by bclothier** [reply]
Just in case, I threw in some more folders & files and made 380MB `.rar` file. Extracted without any problem. Will have to check with @wqweto whether there's something else or if it's a change in 7-zip's internal implementation between back & now.

**Reply by wqweto** [reply]
> Unless 7-zip changed how it handled multi-threaded I/O, it looks like it's running on a single thread in spite of reports of this not working on VB6's version.

Could be this or multi-threading being used upon compression or with other formats. It's been awhile since I abandoned the VszLib (still using it for .7z files in production though)
